File: jump-ii.java
// Time Complexity :O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no
class Solution {
    public int jump(int[] nums) {
        int n = nums.length;
        int currInterval = nums[0];
        int nextInterval = nums[0];
        
        int jumps=1;
        
        if(n<2){
            return 0;
        }
        for(int i=1;i<n;i++){
            nextInterval = Math.max(nextInterval, nums[i]+i);
            
            if(i!=n-1 && i==currInterval){
                jumps++;
                currInterval = nextInterval;
            }
        }
        
        return jumps;
    }
}

File: jumpII.py
'''
Time Complexity --> O(n)
Space Complexity --> O(n)
We can use a BFS approach here
'''
class Solution:
    def jump(self, nums: List[int]) -> int:
        q = []
        level = 1
        q.append(0)
        visited = {}
        n = len(nums)
        if n==1:
            return 0
        while q:
            #print(q)
            s =len(q)
            
            #visited[index]=1
            for i in range(s):
                index = q.pop(0)
                for i in range(1,nums[index]+1):
                    new_index = i+index
                    if new_index not in visited:
                        if new_index == n-1:
                            return level
                        visited[new_index]=1
                        q.append(new_index)
            level+=1
        return level



            
        


            # print(q,end =' ')
            # print(level,end =' ')
            # print(visited)
        #return level
        
